File: src/shared/types/agent.ts
export type LLMProvider = "anthropic" | "openai" | "ollama" | "google" | "vertexai";

export interface LLMConfig {
  provider: LLMProvider;
  model: string;
  apiKey?: string;
  baseUrl?: string;
}

// -- Content Part types (ordered timeline) ---------------------------

export type NotifierType =
  | "state_change"
  | "graph_transition"
  | "sub_graph"
  | "agent_start"
  | "agent_end"
  | "error"
  | "warning";

export interface ThinkingPart {
  type: "thinking";
  id: string;
  content: string;
  isStreaming: boolean;
}

export interface TextPart {
  type: "text";
  id: string;
  content: string;
}

export interface ToolCallPart {
  type: "tool_call";
  id: string;
  toolCallId: string;
  name: string;
  args: Record<string, unknown>;
  result?: string;
  status: "pending" | "running" | "success" | "error";
  startedAt: string;
  completedAt?: string;
}

export interface NotifierPart {
  type: "notifier";
  id: string;
  notifierType: NotifierType;
  label: string;
  detail?: string;
  timestamp: string;
}

export interface AgentStepPart {
  type: "agent_step";
  id: string;
  stepName: string;
  detail: string;
  status: "running" | "completed";
  startedAt: string;
  completedAt?: string;
  subSteps: string[];
}

export type ContentPart = ThinkingPart | TextPart | ToolCallPart | NotifierPart | AgentStepPart;

// -- Agent activity types (used by activity-stream) ------------------

export interface ToolActivity {
  id: string;
  type: "tool";
  name: string;
  args: Record<string, unknown>;
  result?: string;
}

export interface ThinkingActivity {
  id: string;
  type: "thinking";
  text: string;
}

export type AgentActivity = ToolActivity | ThinkingActivity;

// -- Chat message ----------------------------------------------------

export interface ChatMessage {
  id: string;
  conversationId: string;
  role: "user" | "assistant";
  content: string;
  thoughts: string | null;
  toolCalls: ToolCallInfo[] | null;
  createdAt: string;
}

export interface ToolCallInfo {
  id?: string;
  name: string;
  args: Record<string, unknown>;
  result?: string;
}

export interface Conversation {
  id: string;
  title: string;
  createdAt: string;
  updatedAt: string;
}

export type OutputCategory =
  | "explanation"
  | "decision"
  | "code_result"
  | "summary"
  | "response"; // fallback

export type StreamModeName = "updates" | "messages" | "custom" | "values" | "debug";

/** Classifica o tipo de node do LangGraph */
export enum NodeCategory {
  /** Invocação direta do LLM (agent, model) */
  LLM_INVOKE = "LLM_INVOKE",
  /** Execução de tools (tools, tool_executor) */
  TOOL_EXECUTION = "TOOL_EXECUTION",
  /** Subgraph de outro agente (formato "agentName:nodeName") */
  SUBGRAPH = "SUBGRAPH",
  /** Nó interno/middleware — não deve aparecer no frontend */
  INTERNAL = "INTERNAL",
  /** Nó customizado desconhecido */
  UNKNOWN = "UNKNOWN",
}

/** Metadados de um node do LangGraph */
export interface NodeMeta {
  name: string;
  category: NodeCategory;
  label: string;
  isStreamable: boolean;
  subgraphPath?: string[];
}

export type StreamEventType =
  | "thought"
  | "tool_call"
  | "tool_result"
  | "response"
  | "step"
  | "agent_step"
  | "done"
  | "error"
  | "notifier";

export interface StreamEvent {
  id: string;
  seq: number;
  type: StreamEventType;
  mode: StreamModeName;
  data: string;
  meta?: {
    runId?: string;
    parentRunId?: string;
    node?: string;
    nodeCategory?: string;
    toolCallId?: string;
    provider?: LLMProvider;
    model?: string;
    status?: "start" | "update" | "end";
    path?: string[];
    // Turn filter & ordering
    turnRunId?: string;
    insertBefore?: string;
    /** Stable marker ID emitted on the first response part of a turn.
     *  The frontend stores the mapping marker → TextPart ID for insertBefore resolution. */
    firstResponseMarker?: string;
    // Structured output
    category?: OutputCategory;
  };
}

File: python/omnimind_backend/schemas/agent.py
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from omnimind_backend.schemas.common import LLMProvider


class AgentChatRequest(BaseModel):
    message: str = Field(min_length=1, max_length=100000)
    provider: LLMProvider | None = None
    model: str | None = None
    debugSteps: bool = False
    orchestrate: bool = False


StreamModeName = Literal["updates", "messages", "custom", "values", "debug"]
StreamEventType = Literal[
    "thought",
    "tool_call",
    "tool_result",
    "response",
    "step",
    "agent_step",
    "done",
    "error",
    "notifier",
]


class StreamEventMeta(BaseModel):
    runId: str | None = None
    parentRunId: str | None = None
    node: str | None = None
    nodeLabel: str | None = None
    nodeCategory: str | None = None
    userVisible: bool | None = None
    toolCallId: str | None = None
    provider: LLMProvider | None = None
    model: str | None = None
    status: Literal["start", "update", "end"] | None = None
    path: list[str] | None = None
    turnRunId: str | None = None
    insertBefore: str | None = None
    firstResponseMarker: str | None = None
    category: Literal["explanation", "decision", "code_result", "summary", "response"] | None = None


class StreamEvent(BaseModel):
    id: str
    seq: int
    type: StreamEventType
    mode: StreamModeName
    data: str
    meta: StreamEventMeta | None = None


class LogEntry(StreamEvent):
    turnId: str
    wallTime: str


class AgentChatStreamPayload(BaseModel):
    """Internal normalized stream payload from runtime."""

    model_config = ConfigDict(arbitrary_types_allowed=True)

    event: StreamEvent
    assistant_text_delta: str = ""
    assistant_thought_delta: str = ""
    tool_call: dict[str, Any] | None = None
